### ISSUE TITLE
Host switching for multiple HID-devices

### ISSUE THREAD

**Issue by jeblad** [pwr-Solaar/Solaar#870]
**Information**

```
john@hydra:~$ solaar --version
solaar 1.0.3rc1
```

**Is your feature request related to a problem? Please describe.**

I have a background in a somewhat security-minded environment, and noticed that having HID-devices attached to several machines created a security issue, where you by accident can type or copy-paste into machines or windows. You will typically use visual memory for navigation between machines, often by tracking the pointer with your eyes, and then easily forget where your other HID-devices are located. When you type your keyboard are attached somewhere else and hitting [enter] your precious password may easily end up at Twitter. And that is pretty bad.

**Describe the solution you'd like**

It would be a lot better if any HID-device that changes state to a new host also trigger a change for other devices within the HID-group. The obvious problem is how to detect what is the new host. Perhaps it is easy, I don't know.

If you have a mouse and/or keybord with buttons for the different hosts, then hit one of the buttons and both HID-devices should attach to the new host. It should not be necessary to change the host manually for both (or all) devices.

**Describe alternatives you've considered**

One simple alternative is to detach the keyboard from the host if the mouse gets detached. That removes the risk of leaking secrets when the keyboard isn't attached to the same host.

**Additional context**

Moving the pointer to the edge of the screen as Logitech does, is not a perfect solution, as a screen is often repurposed and the physical layout thus does not reflect the virtual layout.

**Comment by pfps** [comment]
As stated, this is almost certainly outside the scope of Solaar, as Solaar would have to repurpose a key on the keyboard and then listen to that key.  

What has been done (@barolo did you do this) is to set up a Linux command that uses Solaar to switch both devices to a different host at the same time using the CLI of Solaar.

**Comment by jeblad** [comment]
Not sure if you know, but Logitech has made both mouse and keyboard with such buttons, and has done so for several years. In particular [M585](https://www.logitech.com/en-us/product/m585-wireless-mouse), [M590](https://www.logitech.com/en-us/product/m590-silent-wireless-mouse) and [MX Master](https://www.logitech.com/en-us/product/mx-master-3.910-005620.html) has a button to switch between hosts, and [MX keys](https://www.logitech.com/en-us/product/mx-keys-wireless-keyboard.920-009294.html) has such keys too. In the case of MX Master and MX Keys they call it Easy-Switch™ button(s).

Read the links, learn something new! ;)

**Comment by barolo** [comment]
@jeblad 
You can use such command to switch both [ or how many you wish ] devices at once:
`solaar config "mx master 2s" change-host "2" && ./bin/solaar config "craft" change-host`

Do it via cli or put it in a script, and launch it via your preferred method [ I have it setup to execute when I reach left border of the screen ]

**Comment by pfps** [comment]
@jeblad  For devices with these buttons Solaar would have to arrange for them to emit an HID++ message (which I believe that they do not normally do), listen for this message, and perform the switch.  The listening is outside the current scope of Solaar.  The solution by @barolo is within scope.

**Comment by jeblad** [comment]
There are at least two different issues here. One is how to move between screens, the other is how to keep the system consistent. A working solution will have to consider both of them. In particular, it is not always possible to use the pointer to “drag” the keyboard into another machine as not all machines will have a graphical user interface.

Only viable option as I see it is to use the mouse host button or keyboard host keys to switch between hosts, and then “drag” the other devices to the new host. That work unless the other devices power up too late to be moved, but in that case they should be blocked due to inconsistent system configuration until they are moved to the current active HID-group, or the group is back to the same host.

The information about what goes on exists inside Solaar. It is the only place where it is possible to maintain consistency among HID devices, and without consistency the HID devices becomes a security issue.

**Comment by pfps** [comment]
If there is to be a solution completely inside Solaar then someone with the appropriate hardware is going to have to take the lead on implementation.

**Comment by barolo** [comment]
@jeblad I gave you a CLI command that can move any amount of devices at once, doing it visually is optional. I don't understand your issue with that

**Comment by pfps** [comment]
Closing for now.   If you are interested in figuring out how to add this functionality to Solaar you can reopen this issue and ask for more information on how to modify Solaar.

**Comment by pfps** [comment]
One significant problem with switching multiple devices is figuring out how to switch.   Some devices keep track of information about the multiple hosts but others don't.  How can Solaar figure out that switching a mouse to host 2 actually switches it to your Mac (as opposed to switching to your Windows PC)?  The solution by @barolo puts the user in the loop, as the user can determine which host number is which host for each device.

**Comment by pfps** [comment]
Closing as probably impossible for Solaar to do.

**Comment by barolo** [comment]
@pfps furthermore, second device then can switch to third, can be unpaired, paired with something else, impossible to keep track of

**Comment by jeblad** [comment]
Quite frankly, this is nonsense! There are two different arguments where one is where the switching is triggered from a device, and in this case the claim is that it is impossible to do. In the other case the switching is triggered from a script, and in this case the claim is that this is possible to do. The device report state changes, and as far as I can tell the buttons can even be reprogrammed.

Quite frankly, your two claims are inconsistent!

**Comment by pfps** [comment]
I await an implementation demonstrating that this can be done in Solaar without the user being involved in keeping track of which host number is which host.

**Comment by barolo** [comment]
@pfps I'm not sure if I understand correctly what he wants any-longer but I think that he's asking for a "Flow" functionality from Logitech's software, it can tie mouse to keyboard [ or vice versa, don't remember ] and tracks them if the software is installed on all participating computers, of course reverse engineering that would be a major undertaking and would require porting solaar to multiple OSes or be faking/interoperable with Logitech's software.
In other words, highly improbable.

**Comment by Flacedoo** [comment]
(Partially related): Does anyone of you know if there's a possibility to extract just this feature for macOS as well? I have a setup with a linux PC where @barolo's solution works like a charme to switch my MX Keys together with my MX Master3. Now I would like to have some kind of script on macOS who does the same. Ideas?

**Comment by barolo** [comment]
@Flacedoo If I remember correctly you can tie devices together via Logitech's software so that they switch together. No idea about scripting though. Solaar goes beyond what's possible via LO in some areas

**Comment by pfps** [comment]
@Flacedoo It would certainly be possible to write a small program to do this.  If you only want it to work for one device the code would be simpler than requiring it to work for different kinds of devices.  The problem is figuring out how to write HID++ commands to the device under MacOS.  That requires someone who knows something about how input devices work in MacOS.

**Comment by Flacedoo** [comment]
Thanks guys for the fast response!
@barolo At least I can't find any option like the in the Logitech Options tool nor the unifying tool. Do you remember where I could find it? As it's heavily unintuitive I could have missed it.
@pfps Do you think [this](https://github.com/libusb/hidapi) could do the job? I've got zero experience with HID device programming but am willing to get my hands dirty with some scripting.

**Comment by pfps** [comment]
@Flacedoo That might work.  It's a C library to someone would have to figure out how to get it running under MacOS and write a little program that sends the right output.   
Solaar does have a program to write HID++ commands to devices (I forgot about it).   It might be easy to port this over to the Mac.  
But someone is going to have to do the work either way.

**Comment by Flacedoo** [comment]
Getting it running on macOS seems already implemented and described, I think I can do that part by myself. As I'm completly new in this project, where can I find which HID commands I would have to send to change the host? Can I simply verbose that from Solaar? Didn't found the exact part in the code..

**Comment by pfps** [comment]
HID2.0++ commands are indirected using information from the device.  So first you need to find out which command number is used for host switching.  Run `solaar show` and look for a line like
```
        DD: CHANGE HOST            {1814}   
```
DD is the command number for the change host function for your device.  Then you need to find out what change host commands look like.  This is in the lib/logitech_receiver/settings_templates.py file in Solaar (but it is hard to tease out of the code if you don't know what is going on).   Logitech functions can have several sub-functions and to change the host the sub-function is 0x10.  The host number is then just the next byte.  Then you need to know how to construct an HID++ 2.0 command from this.  The syntax of these commands is a one-byte length code (use 0x11 for Bluetooth devices), followed by a one-byte device code (use 0x00 for Bluetooth devices), followed by the command number (for my MX Vertical the command number is 12 so I would use 0x0C), then the subfunction (but you should put in some non-zero bits in the lower half of this byte), then the host number to switch to, then padding to fill out to 20 bytes, as in
```
11 00 0C 1A 01 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
```
to switch an MX Vertical to host 2.

You can see the command that Solaar sends by running `solaar -dd config Master change-host 1`.   The last line of output contains the HID++ command sent to the device.  (This assumes that you only have one device whose name includes 'Master'.)
